sqlglot/optimizer.py

```diff
@@ -0,0 +1,27 @@
+import sqlglot.expressions as exp
+
+
+def qualify_tables(expression, db=None, catalog=None):
+    def qualify(node):
+        if isinstance(node, exp.Table):
+            if not node.args.get("db"):
+                node.set("db", db)
+            if not node.args.get("catalog"):
+                node.set("catalog", catalog)
+            return node
+        return node
+
+    return expression.transform(qualify)
+
+
+def qualify_columns(expression, _schema):
+    return expression
+
+
+def optimize(expression, schema=None, db=None, catalog=None):
```

Comment:
I am not entirely clear how qualifying tables is an optimization

Reply:
I think this is a _pre_ optimization step.

Reply:
right, there are a series of steps needed to take every query into canonical form, this is the first rule